### PR TITLE
Increase type safety

### PR DESCRIPTION
**Pull request by valentinpalkovic**
## What Changed

Added some types and removed obsolete "any" type assertions by replacing them through stronger types

## Change Type

Indicate the type of change your pull request is:

- [x] `maintenance`
- [ ] `documentation`
- [ ] `patch`
- [ ] `minor`
- [ ] `major`

<details>
  <summary>📦 Published PR as canary version: <code>0.5.11--canary.89.28723b8.0</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install eslint-plugin-storybook@0.5.11--canary.89.28723b8.0
  # or 
  yarn add eslint-plugin-storybook@0.5.11--canary.89.28723b8.0
  ```
</details>

### CODE REVIEW

**Comment by yannbf** [comment]
Hey @valentinpalkovic thank you so much for this contribution, I love this!!

There are still a couple issues in this PR, could you check them? It happens when running `yarn update-all`:


```js
yarn update-all
yarn run v1.22.17
$ yarn update-configs && yarn update-docs
$ ts-node ./tools/update-configs
/Users/yannbraga/open-source/eslint-plugin-storybook/node_modules/ts-node/src/index.ts:750
    return new TSError(diagnosticText, diagnosticCodes);
           ^
TSError: ⨯ Unable to compile TypeScript:
tools/update-configs.ts:8:28 - error TS7016: Could not find a declaration file for module '../.prettierrc'. '/Users/yannbraga/open-source/eslint-plugin-storybook/.prettierrc.js' implicitly has an 'any' type.

8 import prettierConfig from '../.prettierrc'
                             ~~~~~~~~~~~~~~~~
tools/update-configs.ts:10:16 - error TS2665: Invalid module name in augmentation. Module '../.prettierrc' resolves to an untyped module at '/Users/yannbraga/open-source/eslint-plugin-storybook/.prettierrc.js', which cannot be augmented.

10 declare module '../.prettierrc' {}
                  ~~~~~~~~~~~~~~~~
tools/update-configs.ts:27:7 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{ 'import/no-anonymous-default-export': string; }'.
  No index signature with a parameter of type 'string' was found on type '{ 'import/no-anonymous-default-export': string; }'.

27       setting[rule.ruleId] = rule.meta.docs.recommended || 'error'
         ~~~~~~~~~~~~~~~~~~~~
tools/update-configs.ts:43:29 - error TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type '{ csf: null; recommended: null; 'csf-strict': string; }'.
  No index signature with a parameter of type 'string' was found on type '{ csf: null; recommended: null; 'csf-strict': string; }'.

43   const extendsCategoryId = extendsCategories[category.categoryId]
```

**Review by yannbf — CHANGES_REQUESTED** [review]
Also, would you like to edit your author email from your commit? Seems as if it's not set and I think this might end up not adding you as a contributor to this project as a result.

**Comment by yannbf** [comment]
> Hey @yannbf
> 
> I will take care of it directly after The Easter holidays. :) Have a nice holiday time!

Thank you so much! I wish you an incredible holiday!!

**Comment by valentinpalkovic** [comment]
@yannbf Please take another look.

It seems that `ts-node` does a different kind of type checking in the background than `tsc --noEmit` does. It complains about `strict-mode` specific restrictions, although it shouldn't complain. Therefore, I have disabled `type-checking` for `ts-node` and introduced a further step in the CI pipeline to check Typescript types explicitly via `tsc`.

File: lib/rules/prefer-pascal-case.ts [hunk 1]
```diff
@@ -87,8 +81,10 @@ export = createStorybookRule({
                 const scope = context.getScope().childScopes[0]
                 if (scope) {
                   const variable = ASTUtils.findVariable(scope, name)
-                  for (let i = 0; i < variable?.references?.length; i++) {
-                    const ref = variable.references[i]
+                  const index = variable?.references?.length || 0
```
Comment (yannbf) [inline lib/rules/prefer-pascal-case.ts:84]: maybe `index` should be called something else? like `referenceCount`? in the next line `i` refers to index so it might be slightly confusing

File: tools/update-rules-list.ts [hunk 2]
```diff
@@ -1,21 +1,26 @@
...
+  categories: string
+]
+export type TRuleListWithoutName = TRulesList extends readonly [string, ...infer TRulesWithoutName]
+  ? TRulesWithoutName
+  : never
+
 const createRuleLink = (ruleName: string) =>
   `[\`storybook/${ruleName}\`](./docs/rules/${ruleName}.md)`
 
-const generateConfigBadges = (recommendedConfig: any) =>
```
Comment (yannbf) [inline tools/update-rules-list.ts:12]: Why was this deleted?

File: lib/utils/index.ts [hunk 3]
```diff
@@ -1,25 +1,30 @@
...
   isVariableDeclaration,
   isVariableDeclarator,
 } from './ast'
+import { RuleContext } from '@typescript-eslint/experimental-utils/dist/ts-eslint'
 
-export const docsUrl = (ruleName: any) =>
+export const docsUrl = (ruleName: string) =>
   `https://github.com/storybookjs/eslint-plugin-storybook/blob/main/docs/rules/${ruleName}.md`
 
-export const isPlayFunction = (node: any) => {
```
Comment (yannbf) [inline lib/utils/index.ts:16]: why was this deleted?

**Review by yannbf — CHANGES_REQUESTED** [review]
Thank you so much for making these changes @valentinpalkovic ! I added a couple more comments. Kinda interesting to know about ts-node, I think I might be experiencing a similar issue in another project for something completely different!

File: lib/utils/index.ts (hunk 3 above)
Comment (valentinpalkovic) [reply lib/utils/index.ts:16]: As far as I remember, it was not used at all.

File: tools/update-rules-list.ts (hunk 2 above)
Comment (valentinpalkovic) [reply tools/update-rules-list.ts:12]: As far as I remember, it was not used at all.

File: lib/rules/use-storybook-expect.ts [hunk 4]
```diff
@@ -4,23 +4,22 @@
...
 
-export = createStorybookRule({
+type TDefaultOptions = {
+  storybookJestPath?: string
+}[]
+
+export = createStorybookRule<TDefaultOptions, string>({
   name: 'use-storybook-expect',
-  defaultOptions: [],
+  defaultOptions: [{ storybookJestPath: '@storybook/jest' }],
```
[the remark this answers is not in the document]
Comment (valentinpalkovic) [reply lib/rules/use-storybook-expect.ts:22]: oh. Good catch! That was actually from a feature request, I was working on. Of course, it shouldn't be part of this PR. I will remove it

File: lib/rules/prefer-pascal-case.ts (hunk 1 above)
Comment (valentinpalkovic) [reply lib/rules/prefer-pascal-case.ts:84]: As far as I remember, it was not used at all. I am unsure whether it is a relic of the past or a draft of a future feature. Of course, I can add it back.

Comment (valentinpalkovic) [reply lib/rules/prefer-pascal-case.ts:84]: Sounds good to me!

**Comment by valentinpalkovic** [comment]
@yannbf Could you take another look? :)

File: lib/utils/index.ts (hunk 3 above)
Comment (yannbf) [reply lib/utils/index.ts:16]: That makes sense. I thought I saw it being used somewhere. Thanks!

File: tools/update-rules-list.ts (hunk 2 above)
Comment (yannbf) [reply tools/update-rules-list.ts:12]: Yup you're right, the idea was to be a draft of something I was planning, but so far it wasn't done so it better off be deleted. Thanks!

**Review by yannbf — APPROVED** [review]
Hey @valentinpalkovic sorry for the delay, I've been sick.

It all looks really good, thank you so much for improving this codebase! Now it's much better for new contributors and I'm sure they'll all appreciate your work. Cheers!

**Comment by valentinpalkovic** [comment]
Hey @yannbf thank you for the review!

If everything looks fine, should I merge it? :) I'm just asking because normally I haven't triggered any merges in storybookjs repos so far and I want to make sure if it is ok to do so.
